Add authentication

bio-boris opened the pull request:
This PR adds Basic Authentication to the Elasticsearch client, allowing the service to connect to secured clusters.

The startup script has been updated to use the new authentication variables, resolving the connection issues when pointed at the elastic-next-spike instance. I've deployed and verified this on the next environment. 

Next Steps:

    Populate a test Elasticsearch instance with actual data + auth to fully verify end-to-end functionality.
    Create a user for elastic

Closes #90

bio-boris committed:
Add authentication (#96)

Add authentication (#96)

bio-boris commented on src/utils/config.py:
```diff
@@ -48,3 +58,31 @@ def init_config():
+        logger.info("Using Basic Authentication for Elasticsearch.")
+        return f"Basic {base64_credentials}"
+    elif auth_token:
+        # Bearer authentication
```
There is no way to test this live right now, as this requires enabling SSL communication to be able to generate API keys.

MrCreosote replied:
so with our current elastic search configuration we can't use auth tokens?

bio-boris replied:
Correct. We would have to create and distribute SSL certs and enable HTTPS. @jbezouska-ANL were there any plans to enable HTTPS on the backend?

bio-boris committed:
Delete .github/workflows/docker-image.yml

MrCreosote commented on src/utils/config.py:
```diff
@@ -13,6 +17,9 @@ def init_config():
     es_url = os.environ.get('ELASTICSEARCH_URL', 'http://localhost:9200').strip('/')
+    es_auth_token = os.environ.get('ELASTICSEARCH_AUTH_TOKEN')
+    es_auth_username = os.environ.get('ELASTICSEARCH_AUTH_USERNAME')
+    es_auth_password = os.environ.get('ELASTICSEARCH_AUTH_PASSWORD')
```
What's the reason for supporting both token and user/pwd?

bio-boris replied:
I figured one day we would support tokens.

MrCreosote replied:
:+1:

MrCreosote commented on src/utils/config.py:
```diff
@@ -48,3 +58,31 @@ def init_config():
+
+    Priority:
+    1. If username and password are set, use Basic authentication
+    2. If only auth token is set, use Bearer authentication
```
Why this precedence?

bio-boris replied:
I picked it arbitrarily.

MrCreosote replied:
My thinking is that tokens are more secure than user/pwd so they should be preferred

bio-boris replied:
I think if we can get tokens working, we can delete the username and password code too, but since they aren't working yet, I just got it working with what we have.

MrCreosote replied:
Might as well leave it if it works and just swap the preference

bio-boris replied:
I looked this up the other day and AI says they aren't more secure. 

```
TL;DR: With One Username Per Service, They're Identical
If you have one username per service, then API keys and username/password are literally the same thing security-wise:
Both are:

Base64 encoded strings in headers
Unique per service (no shared credentials)
Equally vulnerable if intercepted
Equally easy to revoke (just disable that one user/key)
Can have identical RBAC permissions

The only remaining differences are cosmetic:

API Key: Authorization: ApiKey <base64(id:key)>
Basic Auth: Authorization: Basic <base64(username:password)>

Both decode to two strings separated by a colon. Both can be scoped, rotated, and revoked independently.
```

MrCreosote replied:
> Equally easy to revoke (just disable that one user/key)

I don't think that's true, it's easier to revoke a token than change a password

bio-boris replied:
How is it easier? Both will require shutting down and restarting the service?

MrCreosote replied:
I assumed it was easier on the ES side, like how in KBase revoking a token is trivial.

If that's not the case and you're saying that passwords and tokens are completely indistinguishable for this use case then we should just drop token support

bio-boris replied:
I don't know, I haven't tested it out, I was thinking more from the side of what happens to the service. Ok I will remove it.

MrCreosote commented on src/utils/wait_for_service.py:
CodeQL is dumb

MrCreosote commented on tests/unit/utils/test_wait_for_service.py:
```diff
@@ -52,3 +53,53 @@ def test_init_config_invalid_config_url_10_timeout(caplog):
+        # Verify success log message
+        assert 'TestService is online!' in caplog.text
+
+    search2_logger.propagate = False
```
You could try/finally to make sure this happens even if the test throws

bio-boris replied:
done

MrCreosote replied:
:+1:

MrCreosote commented on tests/unit/utils/test_wait_for_service.py:
```diff
@@ -52,3 +53,53 @@ def test_init_config_invalid_config_url_10_timeout(caplog):
+        # Verify requests.get was called with the auth token in headers
+        mock_get.assert_called_once()
+        call_args = mock_get.call_args
+        assert call_args[1]['headers']['Authorization'] == 'Bearer test-token'
```
assert_called_once_with() would make this easier to understand and more readable

bio-boris replied:
done

MrCreosote replied:
:+1:

MrCreosote commented on tests/unit/utils/test_wait_for_service.py:
```diff
@@ -52,3 +53,53 @@ def test_init_config_invalid_config_url_10_timeout(caplog):
+
+
+@patch('src.utils.wait_for_service.requests.get')
+def test_wait_for_service_without_auth_token(mock_get, caplog):
```
Same comments as the test above

bio-boris replied:
done

MrCreosote replied:
:+1:

MrCreosote commented on tests/unit/utils/test_config.py:
```diff
@@ -13,3 +15,83 @@ def test_init_config_invalid_config_url():
         os.environ.pop('GLOBAL_CONFIG_URL')
+
+
+@patch('src.utils.config.config')
```
Having to mock out the config in order to test it points out why config singletons aren't the way to go...

bio-boris replied:
Is there an AI for me to do here?

MrCreosote replied:
No, I'm just whinging

MrCreosote commented on tests/unit/utils/test_config.py:
Aren't there any tests that start ES and connect to it? There don't seem to be any tests that actually show the creds work roundtrip

bio-boris replied:
No, it would require setting up the local es with auth in the docker-compose. We could do that in another PR.

MrCreosote replied:
Ok, I think at least one happy and one sad path test would be worth doing

MrCreosote replied:
A 2nd PR is fine

bio-boris replied:
https://github.com/kbase/search_api2/issues/98

MrCreosote replied:
:+1:

MrCreosote commented on src/es_client/query.py:
```diff
@@ -86,6 +86,11 @@ def search(params, meta):
+    auth_header = get_elasticsearch_auth_header()
+    if auth_header:
+        headers['Authorization'] = auth_header
+
```
This appears to be untested

bio-boris replied:
Should we do this in the next PR where we add auth to the docker compose?

MrCreosote replied:
That's fine

bio-boris committed:
Refactor wait_for_service tests for auth token handling

Refactor tests for wait_for_service to reduce duplication and improve clarity.

Commit:
Fix tests

Commit:
remove comments

Commit:
fix flake

bio-boris commented:
@MrCreosote Can we merge this?

MrCreosote commented:
This one is still open https://github.com/kbase/search_api2/pull/97#discussion_r2441103792

Commit:
Tests

bio-boris committed:
Revise docstring for Elasticsearch auth tests

Updated the docstring to clarify test purpose.

Commit:
update creds

Commit:
update creds

Commit:
update creds

Commit:
Tests

Commit:
Remove token

bio-boris committed:
Update query.py

bio-boris committed:
Update config.py

bio-boris committed:
Remove unused logging setup in config.py

Removed unused logging import and logger initialization.

bio-boris committed:
Remove comment on Elasticsearch auth header calculation

Remove comment about calculating Elasticsearch auth header.

bio-boris committed:
Update Elasticsearch headers to use config for auth

Replaced centralized auth header function with config-based authorization header retrieval.

bio-boris committed:
Update Elasticsearch auth token retrieval method

bio-boris committed:
Clean up imports in __main__.py

Removed unused import of get_elasticsearch_auth_header.

bio-boris committed:
Update auth header retrieval in show_indexes

Refactor authentication header retrieval in show_indexes function.

bio-boris committed:
Update Elasticsearch auth header handling

Refactor Elasticsearch authentication header retrieval.

Commit:
Update timeouts

Commit:
update tests

bio-boris commented on src/es_client/query.py:
```diff
@@ -86,10 +86,14 @@ def search(params, meta):
     params = {'allow_no_indices': 'true'}
 
-    resp = requests.post(url, data=json.dumps(options), params=params, headers=headers)
+    resp = requests.post(url, data=json.dumps(options), params=params, headers=headers, timeout=[120, 600])
```
added timeouts, maybe I should instead add an annotation for bandit to ignore these and not add timeouts, since I don't know what the real timeouts should be?

Commit:
update tests

Commit:
update tests

Commit:
Update timeouts

bio-boris commented:
Closing in favor of https://github.com/kbase/search_api2/pull/100
